Figshare CI tests are failing on master

### Bug description

Some Figshare CI tests are currently failing on `master`. See for example https://github.com/jupyterhub/repo2docker/pull/998

```
2021-01-04T22:53:30.7192977Z =================================== FAILURES ===================================
2021-01-04T22:53:30.7195485Z _________ test_detect_figshare[10.6084/m9.figshare.9782777-expected0] __________
2021-01-04T22:53:30.7196356Z 
2021-01-04T22:53:30.7197351Z test_input = '10.6084/m9.figshare.9782777'
2021-01-04T22:53:30.7199608Z expected = {'article': '9782777', 'host': {'api': 'https://api.figshare.com/v2/articles/', 'download': 'download_url', 'filename': 'name', 'filepath': 'files', ...}, 'version': '1'}
2021-01-04T22:53:30.7200946Z 
2021-01-04T22:53:30.7202019Z     @pytest.mark.parametrize("test_input,expected", test_dois_links)
2021-01-04T22:53:30.7203397Z     def test_detect_figshare(test_input, expected):
2021-01-04T22:53:30.7204497Z >       assert Figshare().detect(test_input) == expected
2021-01-04T22:53:30.7205974Z E       AssertionError: assert None == {'article': '...version': '1'}
2021-01-04T22:53:30.7206852Z E         +None
2021-01-04T22:53:30.7209940Z E         -{'host': {'hostname': ['https://figshare.com/articles/', 'http://figshare.com/articles/', 'https://figshare.com/account/articles/'], 'api': 'https://api.figshare.com/v2/articles/', 'filepath': 'files', 'filename': 'name', 'download': 'download_url'}, 'article': '9782777', 'version': '1'}
2021-01-04T22:53:30.7212168Z 
2021-01-04T22:53:30.7213453Z /home/runner/work/repo2docker/repo2docker/tests/unit/contentproviders/test_figshare.py:85: AssertionError
2021-01-04T22:53:30.7215875Z ________ test_detect_figshare[10.6084/m9.figshare.9782777.v1-expected1] ________
2021-01-04T22:53:30.7216769Z 
2021-01-04T22:53:30.7217788Z test_input = '10.6084/m9.figshare.9782777.v1'
2021-01-04T22:53:30.7220116Z expected = {'article': '9782777', 'host': {'api': 'https://api.figshare.com/v2/articles/', 'download': 'download_url', 'filename': 'name', 'filepath': 'files', ...}, 'version': '1'}
2021-01-04T22:53:30.7221470Z 
2021-01-04T22:53:30.7222565Z     @pytest.mark.parametrize("test_input,expected", test_dois_links)
2021-01-04T22:53:30.7223928Z     def test_detect_figshare(test_input, expected):
2021-01-04T22:53:30.7225038Z >       assert Figshare().detect(test_input) == expected
2021-01-04T22:53:30.7226497Z E       AssertionError: assert None == {'article': '...version': '1'}
2021-01-04T22:53:30.7227387Z E         +None
2021-01-04T22:53:30.7230423Z E         -{'host': {'hostname': ['https://figshare.com/articles/', 'http://figshare.com/articles/', 'https://figshare.com/account/articles/'], 'api': 'https://api.figshare.com/v2/articles/', 'filepath': 'files', 'filename': 'name', 'download': 'download_url'}, 'article': '9782777', 'version': '1'}
2021-01-04T22:53:30.7232646Z 
2021-01-04T22:53:30.7234215Z /home/runner/work/repo2docker/repo2docker/tests/unit/contentproviders/test_figshare.py:85: AssertionError
2021-01-04T22:53:30.7236393Z ________ test_detect_figshare[10.6084/m9.figshare.9782777.v2-expected2] ________
2021-01-04T22:53:30.7237274Z 
2021-01-04T22:53:30.7238287Z test_input = '10.6084/m9.figshare.9782777.v2'
2021-01-04T22:53:30.7240468Z expected = {'article': '9782777', 'host': {'api': 'https://api.figshare.com/v2/articles/', 'download': 'download_url', 'filename': 'name', 'filepath': 'files', ...}, 'version': '2'}
2021-01-04T22:53:30.7241821Z 
2021-01-04T22:53:30.7242897Z     @pytest.mark.parametrize("test_input,expected", test_dois_links)
2021-01-04T22:53:30.7244278Z     def test_detect_figshare(test_input, expected):
2021-01-04T22:53:30.7245383Z >       assert Figshare().detect(test_input) == expected
2021-01-04T22:53:30.7246830Z E       AssertionError: assert None == {'article': '...version': '2'}
2021-01-04T22:53:30.7247700Z E         +None
2021-01-04T22:53:30.7250745Z E         -{'host': {'hostname': ['https://figshare.com/articles/', 'http://figshare.com/articles/', 'https://figshare.com/account/articles/'], 'api': 'https://api.figshare.com/v2/articles/', 'filepath': 'files', 'filename': 'name', 'download': 'download_url'}, 'article': '9782777', 'version': '2'}
2021-01-04T22:53:30.7252965Z 
2021-01-04T22:53:30.7254222Z /home/runner/work/repo2docker/repo2docker/tests/unit/contentproviders/test_figshare.py:85: AssertionError
2021-01-04T22:53:30.7256531Z _ test_detect_figshare[https://doi.org/10.6084/m9.figshare.9782777.v1-expected3] _
2021-01-04T22:53:30.7257555Z 
2021-01-04T22:53:30.7258723Z test_input = 'https://doi.org/10.6084/m9.figshare.9782777.v1'
2021-01-04T22:53:30.7261078Z expected = {'article': '9782777', 'host': {'api': 'https://api.figshare.com/v2/articles/', 'download': 'download_url', 'filename': 'name', 'filepath': 'files', ...}, 'version': '1'}
2021-01-04T22:53:30.7262403Z 
2021-01-04T22:53:30.7263462Z     @pytest.mark.parametrize("test_input,expected", test_dois_links)
2021-01-04T22:53:30.7264832Z     def test_detect_figshare(test_input, expected):
2021-01-04T22:53:30.7265924Z >       assert Figshare().detect(test_input) == expected
2021-01-04T22:53:30.7267377Z E       AssertionError: assert None == {'article': '...version': '1'}
2021-01-04T22:53:30.7268263Z E         +None
2021-01-04T22:53:30.7271291Z E         -{'host': {'hostname': ['https://figshare.com/articles/', 'http://figshare.com/articles/', 'https://figshare.com/account/articles/'], 'api': 'https://api.figshare.com/v2/articles/', 'filepath': 'files', 'filename': 'name', 'download': 'download_url'}, 'article': '9782777', 'version': '1'}
2021-01-04T22:53:30.7274136Z 
2021-01-04T22:53:30.7275417Z /home/runner/work/repo2docker/repo2docker/tests/unit/contentproviders/test_figshare.py:85: AssertionError
2021-01-04T22:53:30.7277849Z _ test_detect_figshare[https://doi.org/10.6084/m9.figshare.9782777.v3-expected4] _
2021-01-04T22:53:30.7278876Z 
2021-01-04T22:53:30.7280063Z test_input = 'https://doi.org/10.6084/m9.figshare.9782777.v3'
2021-01-04T22:53:30.7282497Z expected = {'article': '9782777', 'host': {'api': 'https://api.figshare.com/v2/articles/', 'download': 'download_url', 'filename': 'name', 'filepath': 'files', ...}, 'version': '3'}
2021-01-04T22:53:30.7283847Z 
2021-01-04T22:53:30.7284922Z     @pytest.mark.parametrize("test_input,expected", test_dois_links)
2021-01-04T22:53:30.7286283Z     def test_detect_figshare(test_input, expected):
2021-01-04T22:53:30.7287396Z >       assert Figshare().detect(test_input) == expected
2021-01-04T22:53:30.7288842Z E       AssertionError: assert None == {'article': '...version': '3'}
2021-01-04T22:53:30.7289739Z E         +None
2021-01-04T22:53:30.7292761Z E         -{'host': {'hostname': ['https://figshare.com/articles/', 'http://figshare.com/articles/', 'https://figshare.com/account/articles/'], 'api': 'https://api.figshare.com/v2/articles/', 'filepath': 'files', 'filename': 'name', 'download': 'download_url'}, 'article': '9782777', 'version': '3'}
2021-01-04T22:53:30.7294987Z 
2021-01-04T22:53:30.7296267Z /home/runner/work/repo2docker/repo2docker/tests/unit/contentproviders/test_figshare.py:85: AssertionError
2021-01-04T22:53:30.7297533Z =============================== warnings summary ===============================
2021-01-04T22:53:30.7298064Z tests/unit/test_semver.py: 11 warnings
2021-01-04T22:53:30.7300166Z   /opt/hostedtoolcache/Python/3.8.6/x64/lib/python3.8/site-packages/repo2docker/buildpacks/julia/semver.py:116: DeprecationWarning: Function 'semver.match' is deprecated. Deprecated since version 2.10.0.  This function will be removed in semver 3. Use the respective 'semver.VersionInfo.match' instead.
2021-01-04T22:53:30.7301945Z     return semver.match(v_str, self.constraint_str)
2021-01-04T22:53:30.7302357Z 
2021-01-04T22:53:30.7303129Z -- Docs: https://docs.pytest.org/en/stable/warnings.html
2021-01-04T22:53:30.7303651Z 
```
Full log attached here:
[repo2docker-figshare-failing-20210104.txt](https://github.com/jupyterhub/repo2docker/files/5767161/repo2docker-figshare-failing-20210104.txt)

--- a/repo2docker/contentproviders/figshare.py
+++ b/repo2docker/contentproviders/figshare.py
@@ -39,7 +39,7 @@ class Figshare(DoiProvider):
             }
         ]
 
-    url_regex = re.compile(r"(.*)/articles/([^/]+)/([^/]+)/(\d+)(/)?(\d+)?")
+    url_regex = re.compile(r"(.*)/articles/(code/)?([^/]+)/(\d+)(/)?(\d+)?")
 
     def detect(self, doi, ref=None, extra_args=None):
         """Trigger this provider for things that resolve to a Figshare article"""

--- a/repo2docker/contentproviders/figshare.py
+++ b/repo2docker/contentproviders/figshare.py
@@ -39,7 +39,7 @@ class Figshare(DoiProvider):
             }
         ]
 
-    url_regex = re.compile(r"(.*)/articles/([^/]+)/(\d+)(/)?(\d+)?")
+    url_regex = re.compile(r"(.*)/articles/([^/]+)/([^/]+)/(\d+)(/)?(\d+)?")
 
     def detect(self, doi, ref=None, extra_args=None):
         """Trigger this provider for things that resolve to a Figshare article"""
@@ -53,8 +53,8 @@ class Figshare(DoiProvider):
             if any([url.startswith(s) for s in host["hostname"]]):
                 match = self.url_regex.match(url)
                 if match:
-                    self.article_id = match.groups()[2]
-                    self.article_version = match.groups()[4]
+                    self.article_id = match.groups()[3]
+                    self.article_version = match.groups()[5]
                     if not self.article_version:
                         self.article_version = "1"
                     return {

--- a/repo2docker/contentproviders/figshare.py
+++ b/repo2docker/contentproviders/figshare.py
@@ -39,7 +39,10 @@ class Figshare(DoiProvider):
             }
         ]
 
-    url_regex = re.compile(r"(.*)/articles/(code/)?([^/]+)/(\d+)(/)?(\d+)?")
+    # We may need to add other item types in future, see
+    # https://github.com/jupyterhub/repo2docker/pull/1001#issuecomment-760107436
+    # for a list
+    url_regex = re.compile(r"(.*)/articles/(code/|dataset/)?([^/]+)/(\d+)(/)?(\d+)?")
 
     def detect(self, doi, ref=None, extra_args=None):
         """Trigger this provider for things that resolve to a Figshare article"""

--- a/tests/unit/contentproviders/test_figshare.py
+++ b/tests/unit/contentproviders/test_figshare.py
@@ -50,10 +50,16 @@ test_dois_links = [
     (
         "https://doi.org/10.6084/m9.figshare.9782777.v1",
         {"host": test_fig.hosts[0], "article": "9782777", "version": "1"},
+        # $ curl -sIL https://doi.org/10.6084/m9.figshare.9782777.v1 | grep location
+        # location: https://figshare.com/articles/Binder-ready_openSenseMap_Analysis/9782777/1
+        # location: https://figshare.com/articles/code/Binder-ready_openSenseMap_Analysis/9782777
     ),
     (
         "https://doi.org/10.6084/m9.figshare.9782777.v3",
         {"host": test_fig.hosts[0], "article": "9782777", "version": "3"},
+        # $ curl -sIL https://doi.org/10.6084/m9.figshare.9782777.v3 | grep location
+        # location: https://figshare.com/articles/Binder-ready_openSenseMap_Analysis/9782777/3
+        # location: https://figshare.com/articles/code/Binder-ready_openSenseMap_Analysis/9782777
     ),
     (
         "https://figshare.com/articles/title/97827771234",

--- a/tests/unit/contentproviders/test_figshare.py
+++ b/tests/unit/contentproviders/test_figshare.py
@@ -43,9 +43,13 @@ test_dois_links = [
         "10.6084/m9.figshare.9782777.v1",
         {"host": test_fig.hosts[0], "article": "9782777", "version": "1"},
     ),
-    (
+    pytest.param(
         "10.6084/m9.figshare.9782777.v2",
         {"host": test_fig.hosts[0], "article": "9782777", "version": "2"},
+        # $ curl -sIL https://dx.doi.org/10.6084/m9.figshare.9782777.v2 | grep location
+        # location: https://figshare.com/articles/Binder-ready_openSenseMap_Analysis/9782777/2
+        # location: https://figshare.com/articles/code/Binder-ready_openSenseMap_Analysis/9782777
+        marks=pytest.mark.xfail(reason="Problem with figshare version redirects"),
     ),
     (
         "https://doi.org/10.6084/m9.figshare.9782777.v1",
@@ -54,12 +58,13 @@ test_dois_links = [
         # location: https://figshare.com/articles/Binder-ready_openSenseMap_Analysis/9782777/1
         # location: https://figshare.com/articles/code/Binder-ready_openSenseMap_Analysis/9782777
     ),
-    (
+    pytest.param(
         "https://doi.org/10.6084/m9.figshare.9782777.v3",
         {"host": test_fig.hosts[0], "article": "9782777", "version": "3"},
         # $ curl -sIL https://doi.org/10.6084/m9.figshare.9782777.v3 | grep location
         # location: https://figshare.com/articles/Binder-ready_openSenseMap_Analysis/9782777/3
         # location: https://figshare.com/articles/code/Binder-ready_openSenseMap_Analysis/9782777
+        marks=pytest.mark.xfail(reason="Problem with figshare version redirects"),
     ),
     (
         "https://figshare.com/articles/title/97827771234",

